doc/man/postgis.1
.TH "postgis" "1" "" "" "PostGIS"
.SH "NAME"
.LP
postgis - postgis administration helper

.SH "SYNTAX"
.LP
postgis \fIcommand\fR [\fIarguments\fR]

.SH "DESCRIPTION"
.LP
The postgis script supports various commands to help with
administration of a PostGIS installation.

.SH "USAGE"
.LP
Usage information is printed when the script is invoked with no
arguments or with the \fBhelp\fR argument.

.SH "EXAMPLES"
.LP
An example command to check the status of PostGIS in the
\fBtemplate_postgis\fR database:

# \fBpostgis status template_postgis\fR

.SH "AUTHORS"
.LP
Originally written by Sandro Santilli <strk@kbt.io>.

More information is available at http://postgis.net


NEWS
PostGIS 3.4.0rc1
2023/xx/xx

PostGIS 3.4.0beta2
2023/07/29

Changes since 3.4.0beta1 release

* Bug Fixes *

  - #5444, 3.4.0-beta1 fails to build
        (postgis--TEMPLATED--TO--ANY.sql is missing)
        (Sandro Santilli)
  - #5436, doc compile issue on mingw64/msys2 (Sandro Santilli)
  - #4752, build error on ppc64le (Bas Couwenberg)
  - #5449, Clarify licensing of XSL sheets (Regina Obe)
  - #5452, README.postgis, updates for apt based,
         and also put in new version requirements
         (Jelte Fennema, Regina Obe)

* Breaking Changes *

  - #5447, postgis_restore.pl renamed to postgis_restore
           (Regina Obe)

PostGIS 3.4.0dev
2023/xx/xx

This version requires PostgreSQL 12 or higher, GEOS 3.6 or higher, and Proj 6.1+.
To take advantage of all features, GEOS 3.12+ is needed.
To take advantage of all SFCGAL featurs, SFCGAL 1.4.1+ is needed.

Many thanks to our translation teams, in particular:

Teramoto Ikuhiro (Japanese Team)
Vincent Bre (French Team)


* New Features *
  - #5055, complete manual internationalization (Sandro Santilli)
  - #5052, target version support in postgis_extensions_upgrade
    (Sandro Santilli)
  - #5306, expose version of GEOS at compile time (Sandro Santilli)
  - New install-extension-upgrades command in postgis script (Sandro Santilli)
  - #5257, #5261, #5277, Support changes for PostgreSQL 16 (Regina Obe)
  - #5006, GH705, ST_Transform: Support PROJ pipelines (Robert Coup, Koordinates)
  - #5283, [postgis_topology] RenameTopology (Sandro Santilli)
  - #5286, [postgis_topology] RenameTopoGeometryColumn (Sandro Santilli)
  - GH703, [postgis_raster] Add min/max resampling as options (Christian Schroeder)
  - #5336, [postgis_topology] topogeometry cast to topoelement support (Regina Obe)
  - Allow singleton geometry to be inserted into Geometry(Multi*) columns (Paul Ramsey)
  - GH721, New window-based ST_ClusterWithinWin and ST_ClusterIntersectingWin (Paul Ramsey)
  - #5397, [address_standardizer] debug_standardize_address function (Regina Obe)
  - #5373, ST_LargestEmptyCircle, exposes extra semantics on circle finding
    Requires GEOS 3.9+ (Martin Davis)
  - #5267, ST_Project signature for geometry, and two-point signature (Paul Ramsey)
  - #5267, ST_LineExtend for extending linestrings (Paul Ramsey)
  - New coverage functions ST_CoverageInvalidEdges, ST_CoverageSimplify, ST_CoverageUnion (Paul Ramsey)


* Enhancements *
  - #5194, do not update system catalogs from postgis_extensions_upgrade (Sandro Santilli)
  - #5092, reduce number of upgrade paths installed on system (Sandro Santilli)
  - #635, honour --bindir (and --prefix) configure switch for executables (Sandro Santilli)
  - Honour --mandir (and --prefix) configure switch for man pages install path (Sandro Santilli)
  - Honour --htmldir (and --docdir and --prefix) configure switch for
    html pages install path (Sandro Santilli)
  - [postgis_topology] Speed up check of topology faces without edges (Sandro Santilli)
  - [postgis_topology] Speed up coincident nodes check in topology validation  (Sandro Santilli)
  - GH718, ST_QuantizeCoordinates(): speed-up implementation (Even Rouault)
  - Repair spatial planner stats to use computed selectivity for contains/within queries (Paul Ramsey)
  - GH734, Additional metadata on Proj installation in postgis_proj_version (Paul Ramsey)
  - #5177, allow building tools without PostgreSQL server headers (Sandro Santilli)
  - ST_Project signature for geometry, and two-point signature (Paul Ramsey)
  - #4913, ST_AsSVG support for curve types CircularString, CompoundCurve, MultiCurve,
           and MultiSurface (Regina Obe)
  - #5266, #5460 ST_ClosestPoint, ST_ShortestLine support for geography type
    (MobilityDB Esteban Zimanyi, Maxime Schoemans, Paul Ramsey)

* Breaking Changes *
  - #5229, Drop support for Proj < 6.1 and PG 11 (Regina Obe)
  - #5306, GH734, postgis_full_version() and postgis_proj_version()
    now output more information about proj network configuration
    and data paths. GEOS compile-time version
    also shown if different from run-time (Paul Ramsey, Sandro Santilli)
 - #5447, postgis_restore.pl renamed to postgis_restore

PostGIS 3.3.0
2022/08/26

This version requires PostgreSQL 11 or higher, GEOS 3.6 or higher, and Proj 5.2+.
Additional features are enabled if you are running GEOS 3.9+
ST_MakeValid enhancements with 3.10+, ST_ConcaveHull and several other enhancements with GEOS 3.11+.
Requires SFCGAL 1.4.1+ for ST_AlphaShape and ST_OptimalAlphaShape.

The new --enable-lto flag improves math computations.  This new feature is disabled by default
because on some platforms, causes compilation errors (BSD and MingW64 issues have been raised)

Use below to enable it.

 ./configure --enable-lto flag

 * New features *
  - #5116, Topology export/import scripts (Sandro Santilli)
  - ST_Letters creates geometries that look like letters (Paul Ramsey)
  - #5037, postgis_sfcgal: ST_3DConvexHull (Loïc Bartoletti)
  - postgis_sfcgal: sfcgal_full_version - reports BOOST and CGAL version
    (Loïc Bartoletti)
  - GH659, MARC21/XML, ST_GeomFromMARC21, ST_AsMARC21 (Jim Jones)
  - #5132, GH683, sfcgal: ST_3DUnion aggregate function (Sergei Shoulbakov)
  - #5143, SFCGAL ST_AlphaShape and ST_OptimalAlphaShape
           Requires SFCGAL 1.4.1+ (Loïc Bartoletti)
  - #5162, ST_TriangulatePolygon with GEOS 3.11+ (Paul Ramsey, Martin Davis)
  - #5162, ST_SimplifyPolygonHull with GEOS 3.11+ (Paul Ramsey, Martin Davis)
  - #5183, topology.RemoveUnusedPrimitives (Sandro Santilli)

 * Breaking changes *
  - Drop support for PostgreSQL 9.6 and 10 (Regina Obe)
  - Change output for WKT MULTIPOINT. All points now
    wrapped in parens. (Even Rouault)
  - GH674, geometry validation and fixing is disabled
    for ST_DumpAsPolygons and ST_Polygon so it works faster
    but might produce invalid polygons. (Aliaksandr Kalenik)

 * Enhancements *
  - #2861, Add index on topology.node(containing_face) speeding up

loader/postgis.pl
#!/usr/bin/perl

#
# PostGIS - Spatial Types for PostgreSQL
# http://postgis.net
#
# Copyright (C) 2020-2022 Sandro Santilli <strk@kbt.io>
#
# This is free software; you can redistribute and/or modify it under
# the terms of the GNU General Public Licence. See the COPYING file.
#

use File::Basename;


sub usage
{
	print qq{Usage: $0 <command> [<args>]
Commands:
  help               print this message and exits
  enable <database>  enable PostGIS in given database
  upgrade <database> upgrade PostGIS in given database
  status <database>  print PostGIS status in given database
  install-extension-upgrades [--pg_sharedir <dir>] [--extension <name>] [<from>...]
		Ensure files required to upgrade PostGIS from
		the given version are installed on the system.
		The <from> arguments may be either version numbers
		or PostgreSQL share directories to scan to find available
		ones.
};

}

my %SUPPORTED_EXTENSIONS = (
	'address_standardizer' => 1,
	'address_standardizer_data_us' => 1,
	'postgis' => 1,
	'postgis_raster' => 1,
	'postgis_sfcgal' => 1,
	'postgis_tiger_geocoder' => 1,
	'postgis_topology' => 1
);


sub install_upgrade_from
{
	my ( $SHAREDIR, $EXTENSION, $from ) = @_;

	#print "SHAREDIR: $SHAREDIR\n";
	#print "FROM: $from\n";
	die "Please specify a sharedir and a version to install upgrade support for.\n"
		unless $from;

	# sanify ${from}
	die "'${from}': invalid version, only 3 dot-separated numbers optionally followed by alphanumeric string are allowed\n"
		unless $from =~ /^[0-9]*\.[0-9]*\.[0-9]*[a-z1-9]*/;

	# extension to process
	my %extensions = %SUPPORTED_EXTENSIONS;
	my $glob = "postgis*.control";

	if ( defined($EXTENSION) ) {
		%extensions = (
			$EXTENSION => 1
		);
		$glob = $EXTENSION;
	}

	my $EXTDIR = ${SHAREDIR} . '/extension';

	while (my $cfile = glob("${EXTDIR}/${glob}")) {
		# Do stuff
		#print " CFILE: ${cfile}\n";
		my $extname = basename($cfile, '.control');

		unless ( exists( $extensions{$extname} ) )
		{
			if ( defined($EXTENSION) )  {
				print STDERR "ERROR: glob [${glob}] matched unrequested extension [${extname}]\n";
			} else {
				print STDERR "NOTICE: extension [${extname}] is not a core PostGIS extension\n";
			}
			next;
		}

		#print " EXTENSION: [${extname}]\n";

		die "File ${EXTDIR}/${extname}--TEMPLATED--TO--ANY.sql is missing"
			unless -f "${EXTDIR}/${extname}--TEMPLATED--TO--ANY.sql";

		my $shcmd = "ln -fvs '${extname}--TEMPLATED--TO--ANY.sql' '${EXTDIR}/${extname}--${from}--ANY.sql'";
		#print " CMD: ${shcmd}\n";
		my $rv = system($shcmd);
		if ( $rv ) {
			die "Error encountered running: $shcmd: $!";
		}

	}

	return 0; # success
}

sub install_upgrade_from_available
{
	my ($SHAREDIR, $EXTENSION) = @_;
	my $EXTDIR = ${SHAREDIR} . '/extension';

	#print "EXTDIR: ${EXTDIR}\n";

	opendir(my $d, $EXTDIR) || die "Cannot read ${EXTDIR} directory\n";
	my @files = grep { /^postgis--/ && ! /--.*--/ } readdir($d);
	foreach ( @files )
	{
		m/^postgis--(.*)\.sql/;
		my $ver = $1;
		next if $ver eq 'unpackaged'; # we don't want to install upgrade from unpackaged
		print "Found version $ver\n";
		return $rv if my $rv = install_upgrade_from $SHAREDIR, $EXTENSION, $ver;
		#return $rv if $rv; # first failure aborts all
	}
	closedir($d);

	exit 0; # success
}

sub install_extension_upgrades
{
	my $DEFSHAREDIR = `pg_config --sharedir`;
	chop($DEFSHAREDIR);
	my $SHAREDIR = ${DEFSHAREDIR};
	my $EXTENSION;

	my @ver;

	for ( my $i=0; $i<@_; $i++ )
	{
		if ( $_[$i] eq '--pg_sharedir' )
		{
			$i++ < @_ - 1 || die '--pg_sharedir requires an argument';
			$SHAREDIR = $_[$i];
			die "$SHAREDIR is not a directory" unless -d ${SHAREDIR};
			next;
		}
		elsif ( $_[$i] eq '--extension' )
		{
			$i++ < @_ - 1 || die '--extension requires an argument';
			$EXTENSION = $_[$i];
			die "$EXTENSION is not a supported extension"
				unless ( exists( $SUPPORTED_EXTENSIONS{$EXTENSION} ) );
			next;
		}

		push @ver, $_[$i];
	}

	print "Installation target: ${SHAREDIR} ${EXTENSION}\n";

	if ( 0 == @ver )
	{
		print "No versions given, will scan ${SHAREDIR}\n";
		push @ver, ${SHAREDIR};
	}


	foreach ( @ver )
	{
		my $v = $_;
		if ( -d $v ) {
			install_upgrade_from_available($SHAREDIR, ${EXTENSION}, $v);
		} else {
			install_upgrade_from($SHAREDIR, ${EXTENSION}, $v);
		}
	}

}


sub enable
{
	print "Enable called with args: @_\n";
  my $db = shift;
  die "Please specify a database name" unless "$db";
  die "Enable is not implemented yet";
}

sub upgrade
{
  die "Please specify at least a database name" unless @_;

  foreach my $db (@_)
  {
    print "upgrading db $db\n";
    my $LOG=`cat <<EOF | psql -qXtA ${db}
BEGIN;
UPDATE pg_catalog.pg_extension SET extversion = 'ANY'
 WHERE extname IN (
			'postgis',
			'postgis_raster',
			'postgis_sfcgal',
			'postgis_topology',
			'postgis_tiger_geocoder'
  );
SELECT postgis_extensions_upgrade();
COMMIT;
EOF`;
  }
}

sub status
{
  die "Please specify at least a database name" unless @_;

  foreach my $db (@_)
  {
    my $sql = "
SELECT n.nspname
  FROM pg_namespace n, pg_proc p
  WHERE n.oid = p.pronamespace
    AND p.proname = 'postgis_full_version'
";

		my $SCHEMA=`psql -qXtA ${db} -c "${sql}" 2>&1`;
		chop($SCHEMA);
		if ( $? ne 0 )
		{
			$SCHEMA =~ s/^.*FATAL: *//;
			print "db $db cannot be inspected: $SCHEMA\n";
			next;
		}

    if ( ! "$SCHEMA" )
		{
      print "db $db does not have postgis enabled\n";
      next;
    }

    $sql="SELECT ${SCHEMA}.postgis_full_version()";
    my $FULL_VERSION=`psql -qXtA ${db} -c "${sql}"`;
		#print "FULL_VERSION: ${FULL_VERSION}\n";

    #POSTGIS="3.1.0dev r3.1.0alpha1-3-gfc5392de7"
    my $VERSION="unknown";
    if ( $FULL_VERSION =~ /POSTGIS="([^ ]*).*/ )
		{
			$VERSION=$1;
		}

    my $EXTENSION='';
    if ( $FULL_VERSION =~ /\[EXTENSION\]/ )
		{
      $EXTENSION=" as extension";
    }
    my $NEED_UPGRADE='';
    if ( $FULL_VERSION =~ /need upgrade/ )
		{
      $NEED_UPGRADE=" - NEEDS UPGRADE";
    }

    print "db $db has postgis ${VERSION}${EXTENSION} in schema ${SCHEMA}${NEED_UPGRADE}\n";
  }
}

# Avoid NOTICE messages
$ENV{'PGOPTIONS'} = $ENV{'PGOPTIONS'} . ' --client-min-messages=warning';

my $cmd = shift;
if ( $cmd eq "help" ) {
	usage(STDOUT);
	exit 0;
}
elsif ( $cmd eq "enable" ) {
	exit enable(@ARGV);
}
elsif ( $cmd eq "upgrade" ) {
	exit upgrade(@ARGV);
}
elsif ( $cmd eq "status" ) {
	exit status(@ARGV);
}
elsif ( $cmd eq "install-extension-upgrades" ) {
	exit install_extension_upgrades(@ARGV);
}
else {
	print STDERR "Unrecognized command: $cmd\n";
  usage(STDERR);
  exit 1
}
